aggregator/handler/config.go

```diff
@@ -0,0 +1,153 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package handler
+
+import (
+	"errors"
+	"time"
+
+	"github.com/m3db/m3aggregator/aggregator"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+)
+
+var (
+	errUnknownFlushHandlerType         = errors.New("unknown flush handler type")
+	errNoForwardHandlerConfiguration   = errors.New("no forward handler configuration")
+	errNoBroadcastHandlerConfiguration = errors.New("no broadcast handler configuration")
+)
+
+// FlushHandlerConfiguration contains configuration for flushing metrics.
+type FlushHandlerConfiguration struct {
+	// Flushing handler type.
+	Type string `yaml:"type" validate:"regexp=(^blackhole$|^logging$|^forward$|^broadcast$)"`
+
+	// Forward handler configuration.
+	Forward *forwardHandlerConfiguration `yaml:"forward"`
+
+	// Forward handler configuration.
+	Broadcast *broadcastHandlerConfiguration `yaml:"broadcast"`
+}
+
+// NewHandler creates a new flush handler
+func (c *FlushHandlerConfiguration) NewHandler(
+	iOpts instrument.Options,
+) (aggregator.Handler, error) {
+	scope := iOpts.MetricsScope()
+	switch Type(c.Type) {
+	case BlackholeHandler:
+		return NewBlackholeHandler(), nil
+	case LoggingHandler:
+		scope = scope.SubScope("logging").Tagged(map[string]string{
+			"handler-type": "logging",
+		})
+		return NewLoggingHandler(iOpts.SetMetricsScope(scope)), nil
+	case ForwardHandler:
+		if c.Forward == nil {
+			return nil, errNoForwardHandlerConfiguration
+		}
+		scope = scope.SubScope("forward").Tagged(map[string]string{
+			"handler-type":   "forward",
+			"forward-target": c.Forward.Name,
+		})
+		logger := iOpts.Logger().WithFields(xlog.NewLogField("forward-target", c.Forward.Name))
+		return c.Forward.NewHandler(iOpts.SetMetricsScope(scope).SetLogger(logger))
+	case Broadcast:
+		if c.Broadcast == nil {
+			return nil, errNoBroadcastHandlerConfiguration
+		}
+		scope = scope.SubScope("broadcast").Tagged(map[string]string{
+			"handler-type": "broadcast",
```

Comment:
the value for tag "handler-type" will always be override by the sub handler under the broadcast handler, and the metrics name will be "broadcast.forward.foo", which breaks the current metrics. This is why I separated the configuration in my original code, so when constructing the forward handler under a broadcast handler, it only sets "target" tag but not override the "handler-type" tag. Then we can have "target": "ingester", "handler-type": "broadcast".

We also probably don't want to broadcast to a broadcast handler recursively?